### PR TITLE
Consolidate string variable substitutions

### PR DESCRIPTION
**Pull request by eugener**

### CODE REVIEW

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     * @param template template to use
+     * @param context context to get values from
+     * @return resulting string with keys replaced with corresponding values
+     */
+    public static String substitute(String template, Map<String,String> context) {
+        Objects.requireNonNull(context);
+        return substitute(template,context::get);
+    }
+
+    public static void main(String[] args) {
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: is this needed, or should it go to test?

Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: My mistake ... forgot to remove this

Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: Fixed

Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: Its removed

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 2]
```diff
@@ -186,7 +188,7 @@ private void ensureClibs() throws IOException {
         Triplet target = projectConfiguration.getTargetTriplet();
         Path clibPath = getCLibPath();
         if (!Files.exists(clibPath)) {
-            String url = URL_CLIBS_ZIP.replace("${osarch}", target.getOsArch());
+            String url = Strings.substitute(URL_CLIBS_ZIP, Map.of("osarch",target.getOsArch()));
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: missing white space after comma

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 3]
```diff
@@ -406,8 +408,8 @@ Path getLlcPath() throws IOException {
         answer.add(Constants.REFLECTION_JAVA_FILE);
         if (useJavaFX) {
             answer.add(Constants.REFLECTION_JAVAFX_FILE);
-            answer.add(Constants.REFLECTION_JAVAFX_ARCH_FILE
-                    .replace("${archOs}", suffix));
+            answer.add( Strings.substitute(Constants.REFLECTION_JAVAFX_ARCH_FILE,Map.of("archOs", suffix)));
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: remove white space before `Strings`
missing white space after comma

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 4]
```diff
@@ -423,8 +425,7 @@ Path getLlcPath() throws IOException {
         answer.add(Constants.JNI_JAVA_FILE);
         if (useJavaFX) {
             answer.add(Constants.JNI_JAVAFX_FILE);
-            answer.add(Constants.JNI_JAVAFX_ARCH_FILE
-                    .replace("${archOs}", suffix));
+            answer.add( Strings.substitute(Constants.JNI_JAVAFX_ARCH_FILE, Map.of("archOs", suffix)));
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: remove white space before `Strings`

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 5]
```diff
@@ -497,8 +498,7 @@ private Path createReflectionConfig(String suffix) throws IOException {
 
     private Path createJNIConfig(String suffix) throws IOException {
         Path gvmPath = paths.getGvmPath();
-        Path jniPath = gvmPath.resolve(Constants.JNI_ARCH_FILE
-                .replace("${archOs}", suffix));
+        Path jniPath = gvmPath.resolve( Strings.substitute(Constants.JNI_ARCH_FILE, Map.of("archOs", suffix)));
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Same, remove white space before `Strings`

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 6]
```diff
@@ -308,9 +309,13 @@ private static String getChecksumFileName(Path base, String customPart, String o
...
         Logger.logDebug("Process zip javaStaticSdk, target = "+target);
-        String javaZip = JAVA_STATIC_ZIP
-                .replace("${version}", configuration.getJavaStaticSdkVersion())
-                .replace("${target}", target);
+
+        String javaZip = Strings.substitute( JAVA_STATIC_ZIP, Map.of(
+            "version",  configuration.getJavaStaticSdkVersion(),
+            "target",  target
+        ));
+
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java]: remove extra blank lines

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 7]
```diff
@@ -308,9 +309,13 @@ private static String getChecksumFileName(Path base, String customPart, String o
 
     private void downloadJavaZip(String target, Path substratePath) throws IOException {
         Logger.logDebug("Process zip javaStaticSdk, target = "+target);
-        String javaZip = JAVA_STATIC_ZIP
-                .replace("${version}", configuration.getJavaStaticSdkVersion())
-                .replace("${target}", target);
+
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java]: remove blank line

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 8]
```diff
@@ -308,9 +309,13 @@ private static String getChecksumFileName(Path base, String customPart, String o
...
+        String javaZip = Strings.substitute( JAVA_STATIC_ZIP, Map.of(
+            "version",  configuration.getJavaStaticSdkVersion(),
+            "target",  target
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java]: remove extra white space after commas

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 9]
```diff
@@ -319,9 +324,12 @@ private void downloadJavaZip(String target, Path substratePath) throws IOExcepti
 
     private void downloadJavaFXZip(String osarch, Path substratePath ) throws IOException {
         Logger.logDebug("Process zip javafxStaticSdk");
-        String javafxZip = JAVAFX_STATIC_ZIP
-                .replace("${version}", configuration.getJavafxStaticSdkVersion())
-                .replace("${target}", osarch);
+
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java]: remove blank line

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 10]
```diff
@@ -319,9 +324,12 @@ private void downloadJavaZip(String target, Path substratePath) throws IOExcepti
...
+        String javafxZip = Strings.substitute( JAVAFX_STATIC_ZIP, Map.of(
+            "version",  configuration.getJavafxStaticSdkVersion(),
+            "target",  osarch
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java]: remove extra white space after commas

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        Objects.requireNonNull(context, "Context is required");
+        Pattern pattern = Pattern.compile("\\$\\{(.+?)}");
+        Matcher matcher = pattern.matcher(Objects.requireNonNull(template, "Template is required"));
+        // StringBuilder cannot be used here because Matcher expects StringBuffer
+        StringBuffer buffer = new StringBuffer();
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            if (context.apply(key) != null) {
+                // quote to work properly with $ and {,} signs
+                matcher.appendReplacement(buffer, Matcher.quoteReplacement(context.apply(key)) );
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: remove extra white space before last `)`

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Replaces keys within the template with values using context function
+     * Throws IllegalArgumentException if the key found in the template has no corresponding value
+     * @param template template to use
+     * @param context context to get values from
+     * @return resulting string with keys replaced with corresponding values
+     */
+    public static String substitute(String template, Function<String,String> context) {
+        Objects.requireNonNull(context, "Context is required");
+        Pattern pattern = Pattern.compile("\\$\\{(.+?)}");
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: shouldn't be this pattern a static field?

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        Matcher matcher = pattern.matcher(Objects.requireNonNull(template, "Template is required"));
+        // StringBuilder cannot be used here because Matcher expects StringBuffer
+        StringBuffer buffer = new StringBuffer();
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            if (context.apply(key) != null) {
+                // quote to work properly with $ and {,} signs
+                matcher.appendReplacement(buffer, Matcher.quoteReplacement(context.apply(key)) );
+            } else {
+                throw new IllegalArgumentException( String.format("Key `%s` is not found", key));
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: remove white space before `String`

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Replaces keys within the template with values using context map
+     * Throws IllegalArgumentException if the key found in the template has no corresponding value
+     * @param template template to use
+     * @param context context to get values from
+     * @return resulting string with keys replaced with corresponding values
+     */
+    public static String substitute(String template, Map<String,String> context) {
+        Objects.requireNonNull(context);
+        return substitute(template,context::get);
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: add white space after comma

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
+package com.gluonhq.substrate.util;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+
+import org.junit.jupiter.api.Test;
+//import static org.junit.jupiter.api.Assertions.assertNull;
```
Comment (jperedadnr) [inline src/test/java/com/gluonhq/substrate/util/StringsTests.java]: remove unused import

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
+package com.gluonhq.substrate.util;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+
```
Comment (jperedadnr) [inline src/test/java/com/gluonhq/substrate/util/StringsTests.java]: remove extra blank line

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    void testSubstituteSuccess() {
+        var vars = new HashMap<String,String>();
+        vars.put("var1", "VAR1");
+        vars.put("var2", "VAR2");
+
+        String result = Strings.substitute( "123 ${var1} 456 ${var2} 789", vars::get);
+        assertEquals( "123 VAR1 456 VAR2 789", result );
+
```
Comment (jperedadnr) [inline src/test/java/com/gluonhq/substrate/util/StringsTests.java]: remove extra blank line

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 6 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java]: I think those provide more clarity on steps in the code though

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java (hunk 5 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: done

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java (hunk 4 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Done

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java (hunk 3 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Done

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java (hunk 2 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Done

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 7 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java]: Done

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 8 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java]: Done

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 10 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java]: Done

File: src/main/java/com/gluonhq/substrate/util/Strings.java (hunk 11 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: Done

File: src/main/java/com/gluonhq/substrate/util/Strings.java (hunk 12 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: True - done

File: src/main/java/com/gluonhq/substrate/util/Strings.java (hunk 13 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: Done

File: src/main/java/com/gluonhq/substrate/util/Strings.java (hunk 14 above)
Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: Done

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java (hunk 15 above)
Comment (eugener) [reply src/test/java/com/gluonhq/substrate/util/StringsTests.java]: Done

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java (hunk 16 above)
Comment (eugener) [reply src/test/java/com/gluonhq/substrate/util/StringsTests.java]: Done

File: src/test/java/com/gluonhq/substrate/util/StringsTests.java (hunk 17 above)
Comment (eugener) [reply src/test/java/com/gluonhq/substrate/util/StringsTests.java]: Thats for clarity IMO

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 6 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java]: Sure, but one line should be enough?

File: src/main/java/com/gluonhq/substrate/util/Strings.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class Strings {
+
+    private static Pattern substitutionPattern = Pattern.compile("\\$\\{(.+?)}");
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/Strings.java]: It could be final?

Comment (eugener) [reply src/main/java/com/gluonhq/substrate/util/Strings.java]: done

**Pull request merged**